Adopt dynamicDowncast<> in Editing

annevk opened the pull request:
#### 945e760e7496263ff31ff6427771a7252f8810c9
<pre>
Adopt dynamicDowncast&lt;&gt; in Editing
<a href="https://bugs.webkit.org/show_bug.cgi?id=267222">https://bugs.webkit.org/show_bug.cgi?id=267222</a>

Reviewed by Darin Adler.

This leaves some downcasts as an exercise for someone else.

* Source/WebCore/editing/Editing.cpp:
(WebCore::closestEditablePositionInElementForAbsolutePoint):
(WebCore::enclosingElementWithTag):
(WebCore::enclosingAnchorElement):
(WebCore::canMergeLists):
(WebCore::isRenderedTable):
(WebCore::isEmptyTableCell):
(WebCore::isTabSpanNode):
(WebCore::tabSpanNode):
(WebCore::caretMaxOffset):
(WebCore::lineBreakExistsAtPosition):
(WebCore::areIdenticalElements):
(WebCore::isBlockFlowElement):
(WebCore::visibleImageElementsInRangeWithNonLoadedImages):

Canonical link: <a href="https://commits.webkit.org/272993@main">https://commits.webkit.org/272993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/679395c4270fc3fbd3b9b0db74f89e0030ac566b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36418 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30644 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9728 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29716 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34282 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10613 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30142 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9259 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9366 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30161 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37736 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30683 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30483 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35478 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9500 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33369 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11279 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7808 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10077 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10289 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/editing/Editing.cpp:
```diff
@@ -793,7 +795,7 @@ bool isTabSpanTextNode(const Node* node)
 HTMLSpanElement* tabSpanNode(const Node* node)
 {
-    return isTabSpanTextNode(node) ? downcast<HTMLSpanElement>(node->parentNode()) : nullptr;
+    return isTabSpanTextNode(node) ? uncheckedDowncast<HTMLSpanElement>(node->parentNode()) : nullptr;
```
Why `uncheckedDowncast` instead of `dynamicDowncast` here?

annevk replied:
The checks in `isTabSpanTextNode()` assure us that it should succeed. And we only want to convert the parent to a span when that is true, so I'm not sure how dynamicDowncast would help here.

darinadler replied:
Can change the `is` function to return the downcasted pointer instead. `dynamicTabSpanTextNodeDowncast`

annevk replied:
That would require changes to a few files, but it doesn't seem impossible and might clarify things a bit as well. Will give that a go tomorrow.

darinadler replied:
On reflection, this function, `tabSpanNode`, **is** `dynamicTabSpanTextNodeDowncast` and so it’s about merging the two functions into one.

annevk replied:
I'm a bit out of my depth with various things here, but https://gist.github.com/annevk/d982332762548b532cd94236e2dc48f9 on top of this PR is what compiles. It doesn't seem great that I had to remove `const` and couldn't adopt `RefPtr` though. And also the majority of callers only wants a boolean so...

Shall I just revert the `uncheckedDowncast<>` from this PR and call it a day?

darinadler replied:
Removing `const` is no big deal; it’s not super-meaningful to pass a `const Node*` when you can just call `parentNode` and get a non-const `ContainerNode*` to the parent, for example.

Not being able to add `RefPtr` seems an unrelated problem; would be nice to make progress on that safety project, but not necessary to do this at the same time.

We want to minimize the use of `uncheckedDowncast` for safety reasons, but also of `downcast` for performance reasons. But neither of those needs to hold up what we’re doing here in this patch.

annevk replied:
Thanks for the feedback! I will land this patch minus this `uncheckedDowncast<>` and then when I have some bandwidth again I'll put up the above (gist) patch for review.

annevk committed:
Adopt dynamicDowncast<> in Editing
https://bugs.webkit.org/show_bug.cgi?id=267222

Reviewed by Darin Adler.

This leaves some downcasts as an exercise for someone else.

* Source/WebCore/editing/Editing.cpp:
(WebCore::closestEditablePositionInElementForAbsolutePoint):
(WebCore::enclosingElementWithTag):
(WebCore::enclosingAnchorElement):
(WebCore::canMergeLists):
(WebCore::isRenderedTable):
(WebCore::isEmptyTableCell):
(WebCore::isTabSpanNode):
(WebCore::tabSpanNode):
(WebCore::caretMaxOffset):
(WebCore::lineBreakExistsAtPosition):
(WebCore::areIdenticalElements):
(WebCore::isBlockFlowElement):
(WebCore::visibleImageElementsInRangeWithNonLoadedImages):

Canonical link: https://commits.webkit.org/272993@main
